### DISCUSSION TITLE
pcp_combine reading wrong files for precip summing

### DISCUSSION THREAD

**Discussion by Jeff-Duda** [dtcenter/METplus#1201] [File I/O] [answered]
I'm trying to use pcp_combine (MET v.9.0) to sum hourly precip files (from HRRR output, so in GRIB2 format) into 6-hour increments.

I get some kind of error no matter which method I try to use.

1. Using -sum

```
+ pcp_combine 20190815_03 01 20190815_09 06 HRRR_precip_06h_f06.nc
DEBUG 2: Performing sum command: Init/In_Accum/Valid/Out_Accum Times = 20190815_030000/010000/20190815_090000/060000
DEBUG 2: Searching for 6 files with accumulation times of 010000 to sum to a total accumulation time of 060000.
DEBUG 1: [1] File ./HRRR_slim_t03z_f06.grib2 matches valid time of 20190815_090000
ERROR  :
ERROR  : is_bufr_file() -> unable to read header from input file "./HRRR_slim_t03z_f22.grib2"
ERROR  :
```

The file list shows that, indeed, the file HRRR_slim_t03z_f22.grib2 has zero size:

> -rw-r--r-- 1 Jeffrey.Duda wrfruc  468015 Aug 23 19:07 HRRR_slim_t03z_f00.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc 5650670 Aug 23 19:07 HRRR_slim_t03z_f01.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc 6029265 Aug 23 19:07 HRRR_slim_t03z_f02.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc 6168082 Aug 23 19:07 HRRR_slim_t03z_f03.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc 6450784 Aug 23 19:07 HRRR_slim_t03z_f04.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc 6056840 Aug 23 19:07 HRRR_slim_t03z_f05.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc 6210107 Aug 23 19:07 HRRR_slim_t03z_f06.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc 6431397 Aug 23 19:07 HRRR_slim_t03z_f07.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc 6722489 Aug 23 19:07 HRRR_slim_t03z_f08.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc 6712994 Aug 23 19:07 HRRR_slim_t03z_f09.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc 6819782 Aug 23 19:07 HRRR_slim_t03z_f10.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc 6882421 Aug 23 19:07 HRRR_slim_t03z_f11.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc 6865233 Aug 23 19:07 HRRR_slim_t03z_f12.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc 6943919 Aug 23 19:07 HRRR_slim_t03z_f13.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc 7110251 Aug 23 19:07 HRRR_slim_t03z_f14.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc 7288361 Aug 23 19:07 HRRR_slim_t03z_f15.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc 7580667 Aug 23 19:07 HRRR_slim_t03z_f16.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc 7920907 Aug 23 19:07 HRRR_slim_t03z_f17.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc 8243174 Aug 23 19:07 HRRR_slim_t03z_f18.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc       0 Aug 23 19:07 HRRR_slim_t03z_f19.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc       0 Aug 23 19:07 HRRR_slim_t03z_f20.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc       0 Aug 23 19:07 HRRR_slim_t03z_f21.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc       0 Aug 23 19:07 HRRR_slim_t03z_f22.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc       0 Aug 23 19:07 HRRR_slim_t03z_f23.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc       0 Aug 23 19:07 HRRR_slim_t03z_f24.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc       0 Aug 23 19:07 HRRR_slim_t03z_f25.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc       0 Aug 23 19:07 HRRR_slim_t03z_f26.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc       0 Aug 23 19:07 HRRR_slim_t03z_f27.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc       0 Aug 23 19:07 HRRR_slim_t03z_f28.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc       0 Aug 23 19:07 HRRR_slim_t03z_f29.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc       0 Aug 23 19:07 HRRR_slim_t03z_f30.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc       0 Aug 23 19:07 HRRR_slim_t03z_f31.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc       0 Aug 23 19:07 HRRR_slim_t03z_f32.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc       0 Aug 23 19:07 HRRR_slim_t03z_f33.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc       0 Aug 23 19:07 HRRR_slim_t03z_f34.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc       0 Aug 23 19:07 HRRR_slim_t03z_f35.grib2
-rw-r--r-- 1 Jeffrey.Duda wrfruc       0 Aug 23 19:07 HRRR_slim_t03z_f36.grib2

But why is that particular file even being involved in the summing? I only want HRRR_slim_t03z_f0[1-6].grib2 to be involved, and the valid times of these files are correct (using wgrib2 -VT to check). HRRR_slim_t03z_f22.grib2 doesn't even have a valid time within the requested range!

2. Using -add:
```
+ pcp_combine -add HRRR_slim_t03z_f06.grib2 HRRR_slim_t03z_f05.grib2 HRRR_slim_t03z_f04.grib2 HRRR_slim_t03z_f03.grib2 HRRR_slim_t03z_f02.grib2 HRRR_slim_t03z_f01.grib2 -field APCP HRRR_precip_06h_f06.nc
DEBUG 2: Since the "-field" command line option was used, parsing the command line arguments as a list of files.
DEBUG 2: Performing derivation command (sum) for 6 files.
DEBUG 1: Reading data (APCP) from input file: HRRR_slim_t03z_f06.grib2
ERROR  :
ERROR  : yyerror() -> syntax error in file "/tmp/met_config_64750_0"
ERROR  :
ERROR  :    line   = 2
ERROR  :
ERROR  :    column = 4
ERROR  :
ERROR  :    text   = "(nul)"
ERROR  :
ERROR  :
ERROR  : (nul)
ERROR  : (nul)
ERROR  :
```

Not sure what that syntax error means, but I'm guessing this has nothing to do with that.
>(pygraf) [Jeffrey.Duda@fe2 2019081503]$ cat /tmp/met_config_20160_0
APCP

I'll admit that one potential sticking point in doing this is that the input files have two different APCP arrays in them...one is 1-hour accumulations...the other is running-forecast-total accumulations (see below example):

> (pygraf) [Jeffrey.Duda@fe2 2019081503]$ wgrib2 -v HRRR_slim_t03z_f10.grib2
1:0:d=2019081503:var discipline=0 center=59 local_table=1 parmcat=16 parm=196:entire atmosphere:10 hour fcst:
2:443312:d=2019081503:var discipline=0 center=59 local_table=1 parmcat=16 parm=195:1000 m above ground:10 hour fcst:
3:675608:d=2019081503:var discipline=0 center=59 local_table=1 parmcat=16 parm=198:1000 m above ground:9-10 hour max fcst:
4:984129:d=2019081503:var discipline=0 center=59 local_table=1 parmcat=7 parm=199:5000-2000 m above ground:9-10 hour max fcst:
5:1053294:d=2019081503:var discipline=0 center=59 local_table=1 parmcat=7 parm=199:6000-1000 m above ground:9-10 hour max fcst:
6:1077184:d=2019081503:var discipline=0 center=59 local_table=1 parmcat=7 parm=199:2000-0 m above ground:9-10 hour max fcst:
7:1130066:d=2019081503:var discipline=0 center=59 local_table=1 parmcat=7 parm=199:3000-0 m above ground:9-10 hour max fcst:
8:1191389:d=2019081503:RELV Relative Vorticity [1/s]:2000-0 m above ground:9-10 hour max fcst:
9:3463144:d=2019081503:RELV Relative Vorticity [1/s]:1000-0 m above ground:9-10 hour max fcst:
10:6009627:d=2019081503:TCOLG Total Column Integrate Graupel [kg/m^2]:local level type 200 0:9-10 hour max fcst:
**11:6075213:d=2019081503:APCP Total Precipitation [kg/m^2]:surface:0-10 hour acc fcst:
12:6588007:d=2019081503:APCP Total Precipitation [kg/m^2]:surface:9-10 hour acc fcst:**

So, a few questions. 1) Is there a way to specify record number 12 (the one I actually want) when calling pcp_combine instead of just the array title (APCP)? 2) How to solve the problem using -sum.

Thanks.

**Comment by malloryprow** [comment]
@Jeff-Duda I know you can specify the record number in the level information. This is from the MET documentation:

The “level” entry specifies a level type and value:
ANNN for accumulation interval NNN
ZNNN for vertical level NNN

ZNNN-NNN for a range of vertical levels
PNNN for pressure level NNN in hPa
PNNN-NNN for a range of pressure levels in hPa
LNNN for a generic level type
**RNNN for a specific GRIB record number**

I may have had a similar problem with pcp_combine. I was using "-sum" with "-pcprx f.*\.2019060600", and got an error when it tried to read a file "f024.2019060700.PRATE", though the format of the file did not match the argument given to -pcprx. I fixed it by moving around the files structure differently, so it wasn't a big deal, just odd.

**Comment by j-opatz** [comment, marked as answer]
Thank you for your question @Jeff-Duda. And a big thanks @malloryprow for the answers on this!

@Jeff-Duda, I'll build on what was already said, taking each problem you're having in turn.

The problem you've encountered with pcp_combine reading in files that are not associated with your MET run is due to the design of the tool: pcp_combine is supposed to be flexible enough to read in data files ranging widely in format, so when you ran `pcp_combine 20190815_03 01 20190815_09 06 HRRR_precip_06h_f06.nc` the tool by default attempts to read in all files in the runtime directory (or `-pcpdir` if specified). Since one of the files was unreadable with the provided commands, the tool errors out.

As @malloryprow correctly indicated, this issue can be resolved with the `-pcprx` option, which tells pcp_combine which files to open using regular expressions. Specifically, you can use 

`-pcprx "HRRR_slim_t03z_f0[1-6]"` 

to indicate a range of values for the files you're interested in.

The second problem, the one you encountered when running pcp_combine in -add mode, was due to a syntax error. When using the `-field` option, you need to include a variable field name and the pressure level, as follows:

`-field 'name="APCP"; level="A1";'`

Hopefully that helps set things in the right direction. If there are still issues with this pcp_combine run please let us know with a response here. If the information provided has helped you solve this issue, please let us know by marking one of the comments in this discussion as the answer (there's a button in the lower left corner of each comment box to do so) and I'll mark the discussion as closed.

**Comment by Jeff-Duda** [comment]
Thanks to both of you for the quick and helpful response.

Since Mallory's suggestion would require me to be constantly moving files around, although I understand how that would work, I opted to use j's method. pcp_combine outputs some ambiguous wording regarding the separation between record numbers 11 and 12 in my files which forced me to try specifying `-field` both ways, but I was able to confirm that they both produced the same output. But that may not happen in every possible case.

Example output:

> pcp_combine -add HRRR_slim_t09z_f06.grib2 HRRR_slim_t09z_f05.grib2 HRRR_slim_t09z_f04.grib2 HRRR_slim_t09z_f03.grib2 HRRR_slim_t09z_f02.grib2 HRRR_slim_t09z_f01.grib2 -field 'name="APCP"; level="A1";' HRRR_precip_06h_f06.nc
DEBUG 2: Since the "-field" command line option was used, parsing the command line arguments as a list of files.
DEBUG 2: Performing derivation command (sum) for 6 files.
DEBUG 1: Reading data (name="APCP"; level="A1";) from input file: HRRR_slim_t09z_f06.grib2
DEBUG 1: Reading data (name="APCP"; level="A1";) from input file: HRRR_slim_t09z_f05.grib2
DEBUG 1: Reading data (name="APCP"; level="A1";) from input file: HRRR_slim_t09z_f04.grib2
DEBUG 1: Reading data (name="APCP"; level="A1";) from input file: HRRR_slim_t09z_f03.grib2
DEBUG 1: Reading data (name="APCP"; level="A1";) from input file: HRRR_slim_t09z_f02.grib2
DEBUG 1: Reading data (name="APCP"; level="A1";) from input file: HRRR_slim_t09z_f01.grib2
WARNING:
WARNING: MetGrib2DataFile::data_plane() -> Using the first of 2 matching records for "APCP/A1":
WARNING: record 11 field 1: ipdtmpl[29] = 1, 8, 2, 0, 83, 0, 0, 1, 0, 1, 0, 0, 255, 0, 0, 2019, 8, 16, 10, 0, 0, 1, 0, 1, 2, 1, 1, 255, 0
WARNING: record 12 field 1: ipdtmpl[29] = 1, 8, 2, 0, 83, 0, 0, 1, 0, 1, 0, 0, 255, 0, 0, 2019, 8, 16, 10, 0, 0, 1, 0, 1, 2, 1, 1, 255, 0
WARNING: Try setting the 0-based "GRIB2_ipdtmpl_index" and "GRIB2_ipdtmpl_val" config file options to select one.
WARNING:

One issue that stuck me a bit was the syntax of specifying the field entry; there are both single and double quotes involved. pcp_combine did not like it when I reversed the usages, which I was not expecting.

Finally, I request that this type of field specification be added to the documentation for the pcp_combine tool. I know this information can be found elsewhere within the total MET documentation, but I would never have known where to search for it since pcp_combine was all I was trying to use. And even though I have used MET utilities for many years now, I never remember that you can specify some config file options for the bigger tools (e.g., GridStat, MODE, etc.) on the command line. It would help if a note indicating that this can be done, with a link to a page that shows how to do it, could be added to the documentation.

But I do have the method working now, so thank you!

**Reply by JohnHalleyGotway** [reply]
@Jeff-Duda glad you were able to get these details sorted out. I saw your request to update the documentation. I made the following change in a feature branch I'm working on right now:
<img width="747" alt="Screen Shot 2021-10-11 at 10 47 16 AM" src="https://user-images.githubusercontent.com/21087144/136826427-de01c7f0-4b73-43eb-93b3-1329863c3f0b.png">

Does that look sufficient?

FYI, the way we resolve this discussion is by marking one of the responses as being a the "Answer". Either the person asking the question or one of the respondents can do that. We've also decided that once the poster confirms the answer, we'll "LOCK" the discussion to prevent future posts. That encourages folks to post new questions in new discussions rather than old ones. We've found that we lose track of new posts to old discussions.

**Reply by Jeff-Duda** [reply]
Thanks, John. Looks good. I've marked a response as "the answer"
